### DISCUSSION TITLE
[item01] 클라이언트한테 객체를 감춤으로써 얻는 이점?

### DISCUSSION THREAD

**Discussion by JHZLO** [woowacourse-7th-study/effective-java#6] [질문] [answered]
[Item01](https://github.com/woowacourse-7th-study/effective-java/blob/main/item01/jhzlo.md)

정적 팩토리 메서드에 대해 공부하다보니 `장점 3,4,5의 공통점`으로 다음과 같은 부분이 있었습니다.
```text
클라이언트에게 하위 인스턴스를 감출 수 있다.
```

클라이언트에게 하위 인스턴스를 감춤으로써 얻을 수 있는 이점으로는 뭐가 있을지에 대해 토론해보면 좋을 것 같습니다.

저는 일단 다음과 같은 이점이 있다고 생각듭니다.

```text
결국 실제로 행하는 동작(세부적인 구현)은 하위 인스턴스에서 구현되는 경우가 많은데
이를 감춤으로써 클라이언트가 쉽게 내부구조를 건들 수 없게 되고, 이는 유지보수성과 보안의 측면에서 이점이 있다.
```

**Comment by cueon** [comment, marked as answer]
1. 구체적인 메서드 내용이 작성 되어 있지 않더라도, 이를 호출해서 미리 로직을 만들어둘 수 있다
2. 인터페이스에 대한 구현체를 새로 만들더라도, 기존에 작성된 인터페이스 타입을 리턴하는 메서드의 코드를 수정하지 않아도 된다

위 두 가지 장점이 있을 거 같다고 생각합니다

저는 부모 클래스의 인스턴스 대신 자식 클래스의 인스턴스를 자유롭게 사용할 수 있게 된다고 이해했는데
준형님 생각처럼 클라이언트에게 자식 클래스의 인스턴스를 감출 수 있게 된다는 관점은 생각해보지 못했네요

**Reply by JHZLO** [reply]
오! 잘 정리해주셨네요!!

지금까지의 내용을 정리해보자면,
1. 구체적인 메서드 내용이 작성 되어 있지 않더라도, 이를 호출해서 미리 로직을 만들어둘 수 있다. - OCP
2. 인터페이스에 대한 구현체를 새로 만들더라도, 기존에 작성된 인터페이스 타입을 리턴하는 메서드의 코드를 수정하지 않아도 된다. - OCP
3. 하위 인스턴스를 감춤으로써 데이터의 무결성 보장

결국 상속을 통한 `추상화`와 `캡슐화`의 장점으로 귀결된다는 생각이 드네요.

즉, 클라이언트 입장에서는 "무엇을" 해야하는지에만 집중할 수 있고 "어떻게"는 신경쓰지 않게 만들어 추상화의 수준을 끌어 올리는 방식 중에 하나인 것 같습니다.
